React Revamp: Refs and Memoization lesson (#25501)
TheOdinProject/curriculum: react/more-react-concepts/refs_and_memoization/refs_and_memoization.md

~~~jsx
const handleClick = useMemo(
  () => () => setCount((prevState) => prevState + 1),
  []
);
// or
const memoizedHandleClick = useMemo(() => handleClick, []);
~~~

With `useCallback`, we don't need to do that. It's specifically made for functions:

~~~jsx
import { useCallback } from "react";

// Inside a component
// Without useCallback
const handleClick = () => setCount((prevState) => prevState + 1);
// With useCallback
const handleClick = useCallback(
  () => setCount((prevState) => prevState + 1),
  []
);
// or
const memoizedHandleClick = useCallback(handleClick, []);
~~~

<span id="usememo-or-usecallback"></span>
Yay, there's only one arrow function, and it's simpler to read. There's nothing extra to `useCallback` other than it only memoizes functions. So the main difference between `useMemo` and `useCallback` is just the type of value it returns.

Which one should we use, then? Use `useMemo` for *any* value types, and use `useCallback` specifically for functions. At the end of the day, they both do similar things with a tiny difference, so use whatever you prefer.

> ### Assignment

Author: Thanks, let's go with what you suggested, seems like a great conclusion already to me 🙇

### Introduction

How can we do DOM manipulations that we usually do in vanilla JavaScript? Is it even possible in React because of its declarative nature? How about memoization? As we all know, when a state has been changed, React will try to re-render the component, which means that it will destroy all local variables not controlled by React and re-execute them. This mental model also works for a regular function that you create, as a component is just a function. Since this is the case, wouldn't it slow down the app if we have a very expensive calculation re-executed with each re-render?

### Lesson Overview

This section contains a general overview of topics you will learn in this lesson.

- Explore `useRef` hook and its use cases
- Explain memoization and how `useCallback` and `useMemo` can be used

> ### The useRef hook

Reviewer: ```suggestion
### The `useRef` hook
```

Reviewer 2: note: according to the [style guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md#headings), we don't use code snippets in headers

Reviewer: Deleting the sentence-casing issues. I mistook it with the style guide of a different repository.

> The `useRef` hook provides a way to directly access and interact with DOM elements and Child components in React. They are often used when performing imperative actions or accessing specific elements rendered in the DOM. Refs can also *persist* values throughout the component's lifecycle, meaning that the value of the ref will not be destroyed every time a component re-renders. This is very useful when you want to store a value that you want to persist throughout the component's lifecycle without storing it in a state.

Reviewer: Refs aren't *just* used for storing DOM nodes - in fact, that's one of the lesser used functions that they have. They are used for storing returned IDs from setInterval or setTimeout, date objects, input values that do not need to trigger re-renders, etc. Basically, they act as a catch-all for values that need to be stored but "aren't state".

> The `useRef` hook provides a way to directly access and interact with DOM elements and Child components in React. They are often used when performing imperative actions or accessing specific elements rendered in the DOM. Refs can also *persist* values throughout the component's lifecycle, meaning that the value of the ref will not be destroyed every time a component re-renders. This is very useful when you want to store a value that you want to persist throughout the component's lifecycle without storing it in a state.

Author: > Refs aren't _just_ used for storing DOM nodes - in fact, that's one of the lesser used functions that they have. They are used for storing returned IDs from setInterval or setTimeout, date objects, input values that do not need to trigger re-renders, etc. Basically, they act as a catch-all for values that need to be stored but "aren't state".

I did write in the end:
> Refs can also *persist* values throughout the component's lifecycle, meaning that the value of the ref will not be destroyed every time a component re-renders. This is very useful when you want to store a value that you want to persist throughout the component's lifecycle without storing it in a state.

But the start could be changed yeah 👍 I've mostly based the premise of this lesson on DOM Manipulation, because of the number of people writing code that they would write in Vanilla JS (in the server). So I focused more on DOM Manipulation. Yeah refs aren't always the best thing to go for and it's more on how they wrote their components, etc.

Do you think I should add another section below the DOM manipulation section? @Lofty-Brambles

> The `useRef` hook provides a way to directly access and interact with DOM elements and Child components in React. They are often used when performing imperative actions or accessing specific elements rendered in the DOM. Refs can also *persist* values throughout the component's lifecycle, meaning that the value of the ref will not be destroyed every time a component re-renders. This is very useful when you want to store a value that you want to persist throughout the component's lifecycle without storing it in a state.

Reviewer: I'm specifically concerned about the way refs are introduced. We definitely have a section on non-reactive DOM manipulation, so defining refs as an escape hatch [dumbing it down] in this paragraph, would be my way to go.

> The `useRef` hook provides a way to directly access and interact with DOM elements and Child components in React. They are often used when performing imperative actions or accessing specific elements rendered in the DOM. Refs can also *persist* values throughout the component's lifecycle, meaning that the value of the ref will not be destroyed every time a component re-renders. This is very useful when you want to store a value that you want to persist throughout the component's lifecycle without storing it in a state.

Reviewer 3: I think it's _slightly_ better to introduce refs & the `useRef` hook in context of DOM manipulation. It provides an easy to understand premise and obviously a great use-case. By the end of the lesson, we can say "oh by the way you can do much more with refs" and give some assignment readings. So I'm content with how the lesson is doing it currently, though I do understand where you're coming from @Lofty-Brambles.

> The `useRef` hook provides a way to directly access and interact with DOM elements and Child components in React. They are often used when performing imperative actions or accessing specific elements rendered in the DOM. Refs can also *persist* values throughout the component's lifecycle, meaning that the value of the ref will not be destroyed every time a component re-renders. This is very useful when you want to store a value that you want to persist throughout the component's lifecycle without storing it in a state.

Reviewer: @01zulfi o/ ^^

> The `useRef` hook provides a way to directly access and interact with DOM elements and Child components in React. They are often used when performing imperative actions or accessing specific elements rendered in the DOM. Refs can also *persist* values throughout the component's lifecycle, meaning that the value of the ref will not be destroyed every time a component re-renders. This is very useful when you want to store a value that you want to persist throughout the component's lifecycle without storing it in a state.

Author: Well, it is the *most common* use case which the docs agree to as well. I can agree on changing the first line but it seems like we are re-iterating it in the 2nd paragraph, also I don't think it is a complement, more like an alternative with very strong differences? To which I can refer them to check **only** this table: https://react.dev/learn/referencing-values-with-refs#differences-between-refs-and-state

IMO:

The `useRef` hook offers a valuable capability of managing a value throughout a component's lifecycle that's not needed for rendering. Unlike regular variables that get destroyed on each component re-render, refs allow you to store and "remember" values without relying on state. Unlike state, refs do not trigger re-renders. They're kind of an alternative to `useState` with very important differences. To understand their differences, read **only** the table here [Differences between refs and state](https://react.dev/learn/referencing-values-with-refs#differences-between-refs-and-state). 

They serve multiple purposes such as:
- Storing returned IDs from `setInterval` or `setTimeout`, which you can use *later* for perhaps clearing the timers.
- Storing input values that do not need to trigger re-renders.
- Storing _any_ value without impacting the rendering logic, meaning they aren't necessary for calculating the JSX.
- Providing direct access to DOM elements.
- Facilitating imperative behaviors (e.g. scrolling to an element, focusing on an element, triggering an animation, etc. Basically any kind of DOM manipulation).

By leveraging refs, you can conveniently retain important data throughout the lifespan of a component, ensuring its availability whenever needed. However, treat this hook as an "escape hatch", you won't always need them, use them when you **need** to communicate with external APIs, which is often the case with the browser API.

What do you guys think? @01zulfi @Lofty-Brambles Let me know if you guys still have some suggestions

> The `useRef` hook provides a way to directly access and interact with DOM elements and Child components in React. They are often used when performing imperative actions or accessing specific elements rendered in the DOM. Refs can also *persist* values throughout the component's lifecycle, meaning that the value of the ref will not be destroyed every time a component re-renders. This is very useful when you want to store a value that you want to persist throughout the component's lifecycle without storing it in a state.

Author: Well, it is the *most common* use case which the docs agree to as well. I can agree on changing the first line but it seems like we are re-iterating the same information in the 2nd paragraph, also I don't think it is a complement, more like an alternative with very strong differences? To which I can refer them to check **only** this table: https://react.dev/learn/referencing-values-with-refs#differences-between-refs-and-state

I can rewrite it to this:

The `useRef` hook offers a valuable capability of managing a value throughout a component's lifecycle that's not needed for rendering. Unlike regular variables that get destroyed on each component re-render, refs allow you to store and "remember" values without relying on state. Unlike state, refs do not trigger re-renders. They're kind of an alternative to `useState` with very important differences. To understand their differences, read **only** the table here [Differences between refs and state](https://react.dev/learn/referencing-values-with-refs#differences-between-refs-and-state). 

They serve multiple purposes such as:
- Storing returned IDs from `setInterval` or `setTimeout`, which you can use *later* for perhaps clearing the timers.
- Storing input values that do not need to trigger re-renders.
- Storing _any_ value without impacting the rendering logic, meaning they aren't necessary for calculating the JSX.
- Providing direct access to DOM elements.
- Facilitating imperative behaviors (e.g. scrolling to an element, focusing on an element, triggering an animation, etc. Basically, DOM manipulation).

However, treat this hook as an "escape hatch", you won't always need them, use them when you **need** to communicate with external APIs, which is often the case with the browser API.

What do you guys think? @01zulfi @Lofty-Brambles Let me know if you guys still have some suggestions

> The `useRef` hook provides a way to directly access and interact with DOM elements and Child components in React. They are often used when performing imperative actions or accessing specific elements rendered in the DOM. Refs can also *persist* values throughout the component's lifecycle, meaning that the value of the ref will not be destroyed every time a component re-renders. This is very useful when you want to store a value that you want to persist throughout the component's lifecycle without storing it in a state.

Reviewer 3: Great suggestions @Lofty-Brambles @fcasibu 

I really like the rewritten section, @fcasibu, though I have an objection. Listing out all the possible use cases is going to burden the learner with a lot of information. I don't think we would want that. It's valuable that they discover these use cases on their own during coding & researching. As such, I think the changes @Lofty-Brambles suggested above are sufficient and gets the point across that refs are not only for DOM stuff.

> #### DOM manipulation

Reviewer: ```suggestion
#### Non-reactive DOM Manipulation 
```

Author: Thanks, this is better 👍

Also, remember that `useRef` hook isn't just limited to focusing elements. It can be used for various other DOM manipulation scenarios, such as scrolling to a specific position, measuring the dimensions of an element, triggering animations, and basically *any* DOM manipulation that you've done before with vanilla JavaScript. The possibilities are endless! For example, we can change the `useEffect` in the above snippet to do the following. Change the button's text, and after 2 seconds, change the text back. You **should not** do this and only use `useRef` for non-destructive DOM operations, but just an example:

~~~jsx
useEffect(() => {
  buttonRef.current.focus();
  buttonRef.current.textContent = "Hey, I'm different!";
  let timeout = setTimeout(() => {
    buttonRef.current.textContent = "Click Me!";
  }, 2000);

  return () => {
    clearTimeout(timeout);
  };
}, []);
~~~

The best thing about this is that this will **never** trigger a component re-render!

> Another question that might pop up in your mind is, "Why not just use `querySelector` or other DOM manipulation methods that we've done previously in vanilla JavaScript?" Primarily because of performance and using the hook with React, it's ensured that it will be optimized. We should also not stray outside React and use the provided API to work on their ecosystem.

Author: That's a better way to put it, thanks 👍

We can also see that it's similar to the `useState` hook in that it can store some values. The main difference is that `useRef` creates a mutable reference, allowing you to update its value without triggering a re-render. But, `useState` manages an immutable state that triggers re-renders when updated.

> ### The useMemo hook

Reviewer: ```suggestion
### The `useMemo` hook
```

Author: I believe we're not supposed to add code snippets in the headings

> In all of the examples, I would advise you to use the [Profiler component](https://react.dev/reference/react/Profiler) that is provided in the `react` module. If you want a more interactive alternative, use the `Profiler` in the [React Developer Tools](https://react.dev/learn/react-developer-tools). To measure rendering performance. Note that sometimes you don't need to optimize anything because of how fast things are already. As the famous saying goes in software development:

Reviewer 3: ```suggestion
In all of the examples, we would advise you to use the [Profiler component](https://react.dev/reference/react/Profiler) that is provided in the `react` module. If you want a more interactive alternative, use the `Profiler` in the [React Developer Tools](https://react.dev/learn/react-developer-tools). To measure rendering performance. Note that sometimes you don't need to optimize anything because of how fast things are already. As the famous saying goes in software development:
```

> > Premature optimization is the root of all evil -- The Art of Computer Programming by Donald Knuth

Reviewer 3: Love this

The `useMemo` hook provides a way to add memoization inside our components. It's used to optimize expensive or complex calculations where it caches the result of a function call and stores it to be used later without recalculating it. The memoized value is, however, recalculated *only* when the dependencies of the `useMemo` hook change. And yes, this hook's parameters are the same as the `useEffect` hook you already know. The hook takes in two arguments: a `calculateValue` callback and a `dependencies` array.

> #### Memoizing expensive calculations

Reviewer: ```suggestion
#### Memoizing Expensive Calculations
```

Author: I think this can be considered a sentence right?

> Headings should always use sentence case:

In our previous projects, namely the Shopping Cart Project. You have some logic where you calculate the total price of the products added to the cart. You might or might not have a `Cart` component that functions as a drawer, where the user can open the cart every time they either click on the `Add to Cart` button or the Cart icon in the header.

An example of a `Cart` component:

~~~jsx
function Cart({ products }) {
  const totalPrice = products.reduce(
    (total, product) => total + product.price * product.quantity,
    0
  );

  return (
    <div>
      {/* Some other content in the cart */}
      {/* Products to display */}
      <p>
        Total Price: <strong>${totalPrice}</strong>
      </p>
      {/* Some button to checkout */}
    </div>
  );
}
~~~

> In our `Cart` component, we have the total price of the products calculated directly inside the component. Every time the component is rendered or updated, the calculation is performed from scratch! That doesn't sound good... What if the user has added many products to the cart? Then it will lead to a sluggish user experience.

Reviewer 3: We could be more specific than "many". Maybe "hundreds of thousands" or "millions", since at that number will `useMemo` starts to become evidently useful. What do you think?

```suggestion
In our `Cart` component, we have the total price of the products calculated directly inside the component. Every time the component is rendered or updated, the calculation is performed from scratch! That doesn't sound good... What if the user has added hundreds of thousands of products to the cart? Then it will lead to a sluggish user experience.
```

Author: Sounds good, I wrote this thinking of letting the reader interpret "many", but yeah could be more specific 👍

The `reduce` method iterates over each product and performs multiplication and addition for every item in the cart. This operation becomes *increasingly* time-consuming as the number of products increases.

Now imagine a user who frequently opens/closes the cart. Every time the drawer is opened, the `Cart` component is rendered, executing everything inside the component. This results in unnecessary recomputations of the same value even if the cart's content hasn't changed.

Let's see how we can use `useMemo` to address this:

~~~jsx

> import { useMemo } from "react"; // import the `useMemo` hook in the react module

Reviewer 3: Unnecessary comment

```suggestion
import { useMemo } from "react";
```

function Cart({ products }) {
  const totalPrice = useMemo(() => {
    return products.reduce(
      (total, product) => total + product.price * product.quantity,
      0
    );
  }, [products]);

  return (
    <div>
      {/* Some other content in the cart */}
      {/* Products to display */}
      <p>
        Total Price: <strong>${totalPrice}</strong>
      </p>
      {/* Some button to checkout */}
    </div>
  );
}
~~~

In the example above, we can easily memoize the calculated value by wrapping it in a `useMemo`, as the syntax is pretty much the same as `useEffect` and almost works the same. Where `useMemo` will also *execute* the callback on mount, and on subsequent re-renders, it will only *re-execute* the callback whenever one of the dependencies *changes*. In our case, whenever the `products` prop changes.

This way, whenever a user opens/closes the cart multiple times, it will not recalculate the `totalPrice` and use the cached value as long as`products` did not change.

> You can also do an implicit return if you prefer:
> 
> ~~~jsx
> const totalPrice = useMemo(
>   () =>
>     products.reduce(
>       (total, product) => total + product.price * product.quantity,
>       0
>     ),
>   []
> );
> ~~~

Reviewer 3: I'm not sure if we want this example. The learner could probably figure out implicit return in `useMemo` themselves.
```suggestion
```

> #### Referential equality checks

Reviewer: ```suggestion
#### Referential Equality Checks
```

Author: I believe this can also be considered a sentence, and a sentence case should be used. What do you think?

Reviewer: Deleting the sentence-casing issues. I mistook it with the style guide of a different repository.

For this example, we will use the `Profiler` component in the `react` module to measure the component's performance. We will also introduce `memo`.

Do note that this is just a very basic example. You will encounter a lot of passing of values to other components as prop, components that are very heavy to render.

> ~~~jsx

Reviewer: This example might benefit from a codesandbox embed rather than a code snippet, in my opinion.

Author: That's a great idea 👍

Reviewer 3: Yes, let's do codesandbox

~~~

We don't need to create a new function, but this is just to test these two functions. You can also directly do the following:

~~~jsx
// Syntax might be weird, but just remember that `useMemo` can take any value, and a function is also just a value `() => setCount((prevState) => prevState + 1)`

const handleClick = useMemo(
  // first arrow function is useMemo's callback
  // second arrow function is our function that will be called later, this one is going to be the cached value and what's going to be stored in `handleClick`
  () => () => setCount((prevState) => prevState + 1),
  []
);
~~~

After adding the `useMemo` hook, you might notice that you didn't see any difference at all. It's still very slow! The props shouldn't change anymore, then why? That's because whenever a component's `state` changes. It will also **always** re-render its children. Is there a way to fix this? Yes, there is! React in one of its APIs provides the [memo](https://react.dev/reference/react/memo) function that lets you skip re-rendering a component when its props are unchanged. We can use this `memo` and wrap the `ButtonComponent` in it.

> Here is the complete code. Kindly test both the `handleClick` and `memoizedHandleClick` functions and play around by swapping those two to the `onClick` prop in `ButtonComponent`. You will quickly notice significant changes in performance and the logged object in the console.

Reviewer 3: > You will quickly notice significant changes in performance and the logged object in the console.

I think we could be specific here on what values we are expecting changes in.

    startTime,
    commitTime
  ) => {
    console.log({
      id,
      phase,
      actualDuration,
      baseDuration,
      startTime,
      commitTime,
    });
  };

  return (
    <div>
      <h1>{count}</h1>
      <Profiler id="button-component" onRender={onRender}>
        {/* Swap memoizedHandleClick with handleClick and vice versa */}
        <ButtonComponent onClick={memoizedHandleClick}>
          Click me!
        </ButtonComponent>
      </Profiler>
    </div>
  );
}
~~~

Two scenarios that should happen here:

1. If you've passed `handleClick` and the `ButtonComponent` has a memo. It will still re-render. Referential equality check fails (previous prop is *not equal* to the current prop).
2. If you've passed `memoizedHandleClick` and the `ButtonComponent` has a memo. It will not re-render. Referential equality check passes (previous prop is *equal* to the current prop).

This works with all values that will be passed as a prop. You might see it being used frequently with the Context API.

~~~jsx
const value = useMemo(
  () => ({ someState, someFunction }),
  [someState, someFunction]
);

return <Context.Provider value={value}>{children}</Context.Provider>;
~~~

> ### The useCallback hook

Reviewer: ```suggestion
### The `useCallback` hook
```

The `useCallback` hook provides another way to memoize a value, not just **any** value like `useMemo`. It can only memoize a function. Did you see the previous snippet that we have with memoizing a function reference with `useMemo`?

~~~jsx
const handleClick = useMemo(
  () => () => setCount((prevState) => prevState + 1),
  []
);
// or
const memoizedHandleClick = useMemo(() => handleClick, []);
~~~

With `useCallback`, we don't need to do that. It's specifically made for functions:

~~~jsx
import { useCallback } from "react";

// Inside a component
// Without useCallback
const handleClick = () => setCount((prevState) => prevState + 1);
// With useCallback
const handleClick = useCallback(
  () => setCount((prevState) => prevState + 1),
  []
);
// or
const memoizedHandleClick = useCallback(handleClick, []);
~~~

<a name="usememo-or-usecallback"></a>
Yay, there's only one arrow function, and it's simpler to read. There's nothing extra to `useCallback` other than it only memoizes functions. So the main difference between `useMemo` and `useCallback` is just the type of value it returns.

> Which one should I use, then? Use `useMemo` for *any* value types, and use `useCallback` specifically for functions. At the end of the day, they both do similar things with a tiny difference, so use whatever you prefer.

Reviewer 3: ```suggestion
Which one should we use, then? Use `useMemo` for *any* value types, and use `useCallback` specifically for functions. At the end of the day, they both do similar things with a tiny difference, so use whatever you prefer.
```

### Assignment

<div class="lesson-content__panel" markdown="1">

1. The article [When to useMemo and useCallback](https://kentcdodds.com/blog/usememo-and-usecallback) by Kent C. Dodds further introduces more examples of when to use `useMemo` and `useCallback` and why you shouldn't bother using it.
2. We've only learned about a basic implementation of the `useRef` hook. For more examples about its usage and why we should be wary of using the hook (more on the links they provided in the guide), check out the [interactive guide of the React documentation](https://react.dev/reference/react/useRef) for `useRef`. Try your best to complete all the challenges!
3. The article [useRef instead of querySelector in React](https://meje.dev/blog/useref-not-queryselector) by Caleb Olojo briefly tells some unexpected behaviors when trying to manipulate the DOM directly with DOM manipulation methods and why we should prefer `useRef` over other DOM manipulation methods like `querySelector`. Check it out!

> 

Reviewer: Since the example with DOM manipulation was addressed by the lesson, would an example on the use of refs for storing interval/timeout ids be good in the assignements?
Dan Abhramov has a great article on this: https://overreacted.io/making-setinterval-declarative-with-react-hooks/

Author: That's a great addition, thanks 👍

Reviewer 3: Love this article

</div>

### Knowledge check

This section contains questions for you to check your understanding of this lesson on your own. If you’re having trouble answering a question, click it and review the material it links to.

> - [Why should you prefer useRef hook over other DOM manipulation methods like querySelector?](https://meje.dev/blog/useref-not-queryselector)

Reviewer: This article is already mentioned under the assignment point 3.

Author: This is more for knowledge check since this question is related to the article I've linked above

Reviewer: My bad, again - mistook it for addl. resources.
